File: view_planning/tests/cpp/meta_voxel_test.cpp
#include <gtest/gtest.h>
#include "meta_voxel.h"

using namespace visioncraft;

// Fixture class for setting up common data for MetaVoxel tests
class MetaVoxelTest : public ::testing::Test {
protected:
    // Common test setup variables
    Eigen::Vector3d sample_position;
    octomap::OcTreeKey sample_key;
    float sample_occupancy;
    MetaVoxel meta_voxel;

    MetaVoxelTest() 
        : sample_position(1.0, 2.0, 3.0),
          sample_key{100, 200, 300},
          sample_occupancy(0.5),
          meta_voxel(sample_position, sample_key, sample_occupancy) {}

    // Test helper to check property existence and value
    template <typename T>
    void checkProperty(const MetaVoxel& voxel, const std::string& name, const T& expected_value) {
        ASSERT_TRUE(voxel.hasProperty(name));
        auto actual_value = boost::get<T>(voxel.getProperty(name));
        ASSERT_EQ(actual_value, expected_value);
    }
};

TEST_F(MetaVoxelTest, Constructor) {
    // Ensure that the constructor initializes correctly
    EXPECT_EQ(meta_voxel.getPosition(), sample_position);
    EXPECT_EQ(meta_voxel.getOctomapKey(), sample_key);
    EXPECT_FLOAT_EQ(meta_voxel.getOccupancy(), sample_occupancy);
}

TEST_F(MetaVoxelTest, SetAndGetPosition) {
    // Test retrieving position
    ASSERT_EQ(meta_voxel.getPosition(), sample_position);
}

TEST_F(MetaVoxelTest, SetAndGetOccupancy) {
    // Set and test new occupancy
    float new_occupancy = 0.8;
    meta_voxel.setOccupancy(new_occupancy);
    EXPECT_FLOAT_EQ(meta_voxel.getOccupancy(), new_occupancy);
}

TEST_F(MetaVoxelTest, SetAndGetLogOdds) {
    // Test setting and retrieving log-odds, check occupancy consistency
    float log_odds_value = -0.2;
    meta_voxel.setLogOdds(log_odds_value);
    EXPECT_FLOAT_EQ(meta_voxel.getLogOdds(), log_odds_value);

    float expected_occupancy = 1.0 / (1.0 + std::exp(-log_odds_value));
    EXPECT_NEAR(meta_voxel.getOccupancy(), expected_occupancy, 1e-6);
}

TEST_F(MetaVoxelTest, SetAndGetCustomProperties) {
    // Integer property
    meta_voxel.setProperty("temperature", 25);
    checkProperty(meta_voxel, "temperature", 25);

    // Float property
    meta_voxel.setProperty("pressure", 101.3f);
    checkProperty(meta_voxel, "pressure", 101.3f);

    // String property
    std::string status = "active";
    meta_voxel.setProperty("status", status);
    checkProperty(meta_voxel, "status", status);

    // Vector property
    Eigen::Vector3d velocity(0.5, 1.5, -0.5);
    meta_voxel.setProperty("velocity", velocity);
    ASSERT_TRUE(meta_voxel.hasProperty("velocity"));
    ASSERT_EQ(boost::get<Eigen::Vector3d>(meta_voxel.getProperty("velocity")), velocity);
}

TEST_F(MetaVoxelTest, HasProperty) {
    // Test checking property existence
    meta_voxel.setProperty("temperature", 25);
    ASSERT_TRUE(meta_voxel.hasProperty("temperature"));
    ASSERT_FALSE(meta_voxel.hasProperty("nonexistent_property"));
}

TEST_F(MetaVoxelTest, GetNonExistentPropertyThrows) {
    // Test that retrieving a non-existent property throws an error
    ASSERT_THROW(meta_voxel.getProperty("nonexistent_property"), std::runtime_error);
}

// Main runner for all tests
int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: view_planning/tests/python/meta_voxel_test.py
import unittest
import sys
import os
import numpy as np
import pandas as pd

sys.path.append(os.path.abspath("../../build/python_bindings"))
from visioncraft_py import MetaVoxel

class TestMetaVoxel(unittest.TestCase):
    
    # Initialize a dictionary to track test results
    methods_status = {
        "Function Name": [],
        "Tested": [],
        "Status": []
    }
    
    def setUp(self):
        """Set up a new MetaVoxel instance for each test case."""
        self.sample_position = np.array([1.0, 2.0, 3.0])
        self.sample_key = (100, 200, 300)
        self.sample_occupancy = 0.5
        self.meta_voxel = MetaVoxel(self.sample_position, self.sample_key, self.sample_occupancy)

    def record_test(self, function_name, tested=True, status=True):
        """Record the test results for summary reporting."""
        self.methods_status["Function Name"].append(function_name)
        self.methods_status["Tested"].append("✔️" if tested else "❌")
        self.methods_status["Status"].append("✅" if status else "❌")
    
    def test_constructor(self):
        """Test the constructor of MetaVoxel."""
        print("Running test_constructor...")
        try:
            # Check if the MetaVoxel instance was initialized correctly
            np.testing.assert_allclose(self.meta_voxel.getPosition(), self.sample_position, err_msg="Position does not match")
            self.assertAlmostEqual(self.meta_voxel.getOccupancy(), self.sample_occupancy, msg="Occupancy does not match")
            self.assertEqual(self.meta_voxel.getOctomapKey(), self.sample_key, "Octomap key does not match")
            self.record_test("MetaVoxel Constructor", status=True)
        except AssertionError:
            self.record_test("MetaVoxel Constructor", status=False)

    def test_set_get_position(self):
        """Test getting position of MetaVoxel."""
        print("Running test_set_get_position...")
        try:
            np.testing.assert_allclose(self.meta_voxel.getPosition(), self.sample_position, err_msg="Position does not match")
            self.record_test("getPosition", status=True)
        except AssertionError:
            self.record_test("getPosition", status=False)

    def test_set_get_occupancy(self):
        """Test setting and retrieving occupancy."""
        print("Running test_set_get_occupancy...")
        try:
            self.meta_voxel.setOccupancy(0.8)
            self.assertAlmostEqual(self.meta_voxel.getOccupancy(), 0.8, msg="Occupancy does not match after setOccupancy")
            self.record_test("setOccupancy/getOccupancy", status=True)
        except AssertionError:
            self.record_test("setOccupancy/getOccupancy", status=False)

    def test_set_get_log_odds(self):
        """Test setting and retrieving log-odds."""
        print("Running test_set_get_log_odds...")
        try:
            log_odds_value = -0.2
            self.meta_voxel.setLogOdds(log_odds_value)
            self.assertAlmostEqual(self.meta_voxel.getLogOdds(), log_odds_value, msg="Log-odds does not match after setLogOdds")
            expected_occupancy = 1.0 / (1.0 + np.exp(-log_odds_value))
            self.assertAlmostEqual(self.meta_voxel.getOccupancy(), expected_occupancy, places=2, msg="Occupancy does not match expected value after setting log-odds")
            self.record_test("setLogOdds/getLogOdds", status=True)
        except AssertionError:
            self.record_test("setLogOdds/getLogOdds", status=False)

    def test_set_get_property(self):
        """Test setting and retrieving custom properties."""
        print("Running test_set_get_property...")
        try:
            # Integer property
            self.meta_voxel.setProperty("temperature", 25)
            self.assertEqual(self.meta_voxel.getProperty("temperature"), 25, "Temperature property does not match")
            
            # Float property
            self.meta_voxel.setProperty("pressure", 101.3)
            self.assertAlmostEqual(self.meta_voxel.getProperty("pressure"), 101.3, places=2, msg="Pressure property does not match")
            
            # String property
            self.meta_voxel.setProperty("status", "active")
            self.assertEqual(self.meta_voxel.getProperty("status"), "active", "Status property does not match")
            
            # Vector property
            vector_property = np.array([0.5, 1.5, -0.5])
            self.meta_voxel.setProperty("velocity", vector_property)
            np.testing.assert_allclose(self.meta_voxel.getProperty("velocity"), vector_property, err_msg="Velocity property does not match")
            
            self.record_test("setProperty/getProperty", status=True)
        except AssertionError:
            self.record_test("setProperty/getProperty", status=False)

    def test_has_property(self):
        """Test checking existence of a property."""
        print("Running test_has_property...")
        try:
            self.meta_voxel.setProperty("temperature", 25)
            self.assertTrue(self.meta_voxel.hasProperty("temperature"), "hasProperty failed for existing property")
            self.assertFalse(self.meta_voxel.hasProperty("nonexistent_property"), "hasProperty should return False for non-existent property")
            self.record_test("hasProperty", status=True)
        except AssertionError:
            self.record_test("hasProperty", status=False)

    def test_get_nonexistent_property(self):
        """Test retrieving a non-existent property raises an error."""
        print("Running test_get_nonexistent_property...")
        try:
            with self.assertRaises(RuntimeError, msg="Expected error for non-existent property not raised"):
                self.meta_voxel.getProperty("nonexistent_property")
            self.record_test("getProperty", status=True)
        except AssertionError:
            self.record_test("getProperty", status=False)

    @classmethod
    def tearDownClass(cls):
        """Print the testing summary after all tests are run."""
        print("\nTesting Summary:")
        df = pd.DataFrame(cls.methods_status)
        print(df.to_markdown(index=False))

# Run the tests
if __name__ == '__main__':
    unittest.main()
